File: algorithms/lectures/merge.js


function sort(a) {
  if (a.length < 2) return a; 
  const mid = parseInt(a.length / 2);
  const b = sort(a.slice(0, mid));
  const c = sort(a.slice(mid));
  return merge(b,c)
  
}

function merge(a,b) {
  
  const n = a.length
  const m = b.length
  const c = Array(n+m).fill(null);
  let i = 0;
  let j = 0;
  let k = 0;

  while (i < n || j < m) {
    if (j == m || (i < n && a[i] < b[j])) {
      c[k] = a[i];
      k+=1;
      i+=1;
    } else {
      c[k]= b[j];
      k+=1;
      j+=1;
    }
      
  }
  return c;
}


console.log(sort([23,4,56,3]))

File: algorithms/lectures/merge.py

def sort(a):
    if len(a) < 2:
        return a
    mid = len(a) // 2
    b = sort(a[:mid])
    c = sort(a[mid:])
    print(b,c)
    return merge(b,c)
def merge(a,b):
    n = len(a)
    m = len(b)
    i = j = k = 0
    c = [None] * (n+m)
    while i < n or j < m:
        if j == m or (i < n and a[i] < b[j]):
            c[k] = a[i]
            k +=1
            i +=1
        else:
            c[k] = b[j]
            k +=1
            j +=1
    return c

if __name__ == "__main__":
    a = [23, 1,34, 2, 3012, 32349, 11, 10, 23123]
    print(sort(a))

